app/controllers/article_controller.rb

```diff
@@ -1,89 +1,66 @@
+    @vertical_articles = {}
     @categories.each do |category|
-      articles = category.articles_list(@collection) # The articles for this category
-      @vertical_articles[category] = sort_vertically(articles)
+      current_articles = articles.where(categories: { id: category.id })
+      @vertical_articles[category] = sort_vertically(current_articles)
     end
 
-    # Do the same for uncategorized articles
-    if @uncategorized_articles.present?
-      @uncategorized_articles = sort_vertically(@uncategorized_articles)
-    end
+    @uncategorized_articles = sort_vertically(
+      articles.where(categories: { id: nil }).reorder(:title)
+    )
   end
 
   def delete
-    if @article.link_list.empty? && @article.target_article_links.empty?
-      if @article.created_by_id == current_user.id || current_user.like_owner?(@collection)
-        @article.destroy 
-        redirect_to collection_subjects_path(@collection.owner, @collection)
-      else
-        flash.alert = t('.only_subject_owner_can_delete')
-        redirect_to collection_article_show_path(@collection.owner, @collection, @article.id)
-      end
+    result = Article::Destroy.call(
+      article: @article,
+      user: current_user,
+      collection: @collection
+    )
+
+    if result.success?
+      redirect_to collection_subjects_path(@collection.owner, @collection)
     else
-      flash.alert = t('.must_remove_referring_links')
+      flash.alert = result.message
       redirect_to collection_article_show_path(@collection.owner, @collection, @article.id)
     end
   end
 
   def update
-    old_title = @article.title
-    gis_truncated = gis_truncated?(params[:article], GIS_DECIMAL_PRECISION)
-
-    @article.attributes = article_params
-    if params['save']
-      #process_source_for_article
-      if @article.save
-        if old_title != @article.title
-          rename_article(old_title, @article.title)
-        end
+    if params[:save]
+      result = Article::Update.call(
+        article: @article,
+        article_params: article_params
+      )
+
+      if result.success?
         record_deed
-        flash[:notice] = t('.subject_successfully_updated')
-        if gis_truncated 
-          flash[:notice] << t('.gis_coordinates_truncated', precision: GIS_DECIMAL_PRECISION, count: GIS_DECIMAL_PRECISION)
-        end
-        redirect_to :action => 'edit', :article_id => @article.id
+
+        flash[:notice] = result.notice
+        redirect_to collection_article_edit_path(@collection.owner, @collection, @article)
       else
-        render :action => 'edit'
+        @article = result.article
+        render :edit, status: :unprocessable_entity
       end
-    elsif params['autolink']
+    elsif params[:autolink]
       @article.source_text = autolink(@article.source_text)
+
       flash[:notice] = t('.subjects_auto_linking')
-      redirect_to :action => 'edit', :article_id => @article.id
+      render :edit
```

Comment:
We are doing `redirect_to` here before, which will reload and defeats the purpose of the unpersisted change. We should instead do `render :edit` so the changes apply.

@benwbrum

Reply:
Agreed 100%.